content/shmem_init_thread.tex

```diff
@@ -0,0 +1,78 @@
+\apisummary{
+Initializes the \openshmem{} library, similar to SHMEM\_INIT(), and in addition
+perform the initialization required for thread-safe invocation of \openshmem{} functions.}
+
+\begin{apidefinition}
+
+\begin{Csynopsis}
+int shmem_init_thread(int requested, int *provided);
+\end{Csynopsis}
+
+\begin{Fsynopsis}  
+INTEGER requested, provided
+     
+SHMEM_INIT_THREAD(requested, provided)
+\end{Fsynopsis}  
+
+\begin{apiarguments}
+\apiargument{IN}{requested}  {The thread level support requested by the user.
+The correct values are SHMEM\_THREAD\_SINGLE, SHMEM\_THREAD\_FUNNELED, or SHMEM\_THREAD\_MULTIPLE}
+\apiargument{OUT}{provided}{The thread level support provided by the \openshmem{} implementation.}
+\end{apiarguments}
+
+
+\apidescription{
+    \FUNC{shmem\_init\_thread}  initializes the \openshmem{} library similar to
+    \FUNC{shmem\_init}, and in addition perform the initialization required for
+    thread-safe invocation of \openshmem{} functions. The argument
+    \textit{requested} is used to specify the desired level of thread support.
+    The function returns the support provided by the library. There are four 
+    levels of threading support.
+ 
+ \begin{itemize}
+\item SHMEM\_THREAD\_SINGLE: The \openshmem{} program may not be multithreaded, 
+and all \openshmem{} interfaces are invoked by a single thread. 
+
+\item SHMEM\_THREAD\_FUNNELED: 
+The \openshmem{} program may be multithreaded, however, the 
+program must ensure that only one thread invokes the \openshmem{}
+interfaces.
+
+\item SHMEM\_THREAD\_SERIALIZED: 
+The \openshmem{} program may be multithreaded, however, the 
+program must ensure that only one thread invokes the \openshmem{}
+interfaces at any given instance i.e., the \openshmem{} interfaces 
+are not invoked concurrently by multiple threads.
+
+\item SHMEM\_THREAD\_MULTIPLE: 
+The \openshmem{} program may be multithreaded, and any 
+thread may invoke the \openshmem{} interfaces.
+
+\item SHMEM\_THREAD\_UNSUPPORTED: The \openshmem{} specification 
```

Comment:
> Do you agree that we should support backward comptability, i.e., no thread support ?
> I want to be careful about what is scoped under "backwards compatibility". As of 1.3, using the OpenSHMEM library with threads is undefined. Just because a developer is using this undefined space and the programs are working, does not mean we need to ensure that that program keeps working under 1.4+ without modification.
> 
> I know of two implementations -- Cray and Sandia -- that have some multithreading extensions. Both of these interfaces are of the TLS-oriented, thread-registered variety, which is an API extension without a direct analog to this or the contexts proposal. It is completely up to those implementations to continue to provide these extensions in way compatible with the multithreading proposal here, assuming it is ratified in 1.4.
> 

> If so, I don’t think just having SINGLE or FUNNELED captures that. The implementation can choose to make just SHMEM_INIT and functions that can be called outside the INIT/FINALIZE thread safe, and rest of it not thread safe, which is an useful implementation. But this is not supported by any of one of the four modes.

> I think this proposal should identify and specify some set of OpenSHMEM functions that are always thread safe; e.g., shmem_my_pe() and shmem_n_pes(). (Obviously, this will need discussion.) If we eventually add shmem_is_initialized(), then I think this would be another always-thread-safe function (not to mention, the only function other than two initialization routines that is safe to call before library initialization).

@nspark Ok, I understand your position better now. This was in fact the original ticket (18a6995dace0b3602129157161a0bacfae3480c6). However, in the latest discussion we had on the WG, we thought there was value in supporting the models out there in addition to no thread support.  So, this was trying to achieve that. I’m ok, to go back the previous semantic, if this is what the community wants.